selfdrive/ui/soundd.cc

```diff
@@ -21,67 +21,54 @@ void sigHandler(int s) {
+      {AudibleAlert::CHIME_WARNING2, "warning_2.wav", false},
+      {AudibleAlert::CHIME_WARNING2_REPEAT, "warning_2.wav", true},
+      {AudibleAlert::CHIME_WARNING_REPEAT, "warning_repeat.wav", true},
+      {AudibleAlert::CHIME_ERROR, "error.wav", false},
+      {AudibleAlert::CHIME_PROMPT, "error.wav", false},
     };
     for (auto &[alert, fn, loops] : sound_list) {
       QSoundEffect *s = new QSoundEffect(this);
-      QObject::connect(s, &QSoundEffect::statusChanged, this, &Sound::checkStatus);
-      s->setSource(QUrl::fromLocalFile(fn));
+      QObject::connect(s, &QSoundEffect::statusChanged, [=]() {
+        assert(s->status() != QSoundEffect::Error);
+      });
+      s->setSource(QUrl::fromLocalFile(sound_asset_path + fn));
       sounds[alert] = {s, loops ? QSoundEffect::Infinite : 0};
     }
 
-    sm = new SubMaster({"carState", "controlsState"});
-
     QTimer *timer = new QTimer(this);
     QObject::connect(timer, &QTimer::timeout, this, &Sound::update);
     timer->start();
   };
-  ~Sound() {
-    delete sm;
-  };
-
-private slots:
-  void checkStatus() {
-    QSoundEffect *s = qobject_cast<QSoundEffect*>(sender());
-    assert(s->status() != QSoundEffect::Error);
-  }
 
   void update() {
-    sm->update(100);
-    if (sm->updated("carState")) {
+    sm.update(100);
+    if (sm.updated("carState")) {
       // scale volume with speed
-      volume = util::map_val((*sm)["carState"].getCarState().getVEgo(), 0.f, 20.f,
-                             Hardware::MIN_VOLUME, Hardware::MAX_VOLUME);
+      float volume = util::map_val(sm["carState"].getCarState().getVEgo(), 0.f, 20.f,
+                                   Hardware::MIN_VOLUME, Hardware::MAX_VOLUME);
       for (auto &[s, loops] : sounds) {
         s->setVolume(std::round(100 * volume) / 100);
       }
     }
-    if (sm->updated("controlsState")) {
-      const cereal::ControlsState::Reader &cs = (*sm)["controlsState"].getControlsState();
-      setAlert({QString::fromStdString(cs.getAlertText1()),
-                QString::fromStdString(cs.getAlertText2()),
-                QString::fromStdString(cs.getAlertType()),
-                cs.getAlertSize(), cs.getAlertSound()});
-    } else if (sm->rcv_frame("controlsState") > 0 && (*sm)["controlsState"].getControlsState().getEnabled() &&
-               ((nanos_since_boot() - sm->rcv_time("controlsState")) / 1e9 > CONTROLS_TIMEOUT)) {
-      setAlert(CONTROLS_UNRESPONSIVE_ALERT);
+    if (sm.updated("controlsState")) {
+      setAlert(sm["controlsState"].getControlsState().getAlertSound());
+    } else if (sm.rcv_frame("controlsState") > 0 && sm["controlsState"].getControlsState().getEnabled() &&
+               ((nanos_since_boot() - sm.rcv_time("controlsState")) / 1e9 > CONTROLS_TIMEOUT)) {
+      setAlert(CONTROLS_UNRESPONSIVE_ALERT.sound);
     }
   }
 
-  void setAlert(Alert a) {
-    if (!alert.equal(a)) {
-      alert = a;
+  void setAlert(AudibleAlert sound) {
+    if (current_sound != sound) {
```

Comment:
this isn't exactly right. if two different alerts that have the same sound play one after the other, then this will only play the first

Reply:
hmmm..I didn't realize this problem before

Reply:
would be nice also write some tests for this. we have `selfdrive/ui/tests/test_soundd.py`, but that only runs on comma two